### PR TITLE
Example generation improvement

### PR DESCRIPTION
**Pull request by jianyexi**

### CODE REVIEW

File: lib/commands/generate-examples.ts [hunk 1]
```diff
@@ -24,6 +24,16 @@ export const builder: yargs.CommandBuilder = {
     describe: "the directory path contains payload.",
     string: true,
   },
+  c: {
+    alias: "config",
+    describe: "config path.",
```
Comment (raych1) [inline lib/commands/generate-examples.ts:29]: How about describe it as 'Readme config path'?

File: lib/generator/mocker.ts [hunk 2]
```diff
@@ -12,7 +14,7 @@ export default class Mocker {
       case "array":
         return this.generateArray(paramSpec, paramName, arrItem);
       default:
-        console.log(`Unknown type ${paramSpec.type}.`);
+        log.info(`unknown type ${paramSpec.type}.`);
```
Comment (raych1) [inline lib/generator/mocker.ts:17]: Are you going to update all other logging place to use `log.`?

File: lib/generator/swaggerMocker.ts [hunk 3]
```diff
@@ -128,21 +154,15 @@ export default class SwaggerMocker {
...
-
-  private getSpecByDiscriminatorValue(value: string) {
-    const spec = this.getSpecDefinition(value);
-    if (spec) {
-      return spec;
-    }
-    for (const definition of Object.values(this.spec.definitions)) {
-      if ((definition as any)["x-ms-discriminator-value"] === value) {
-        return definition;
+  private getCache(schema:any) {
```
Comment (raych1) [inline lib/generator/swaggerMocker.ts:157]: Do we need to pass in cacheType when try to get the value from cache?

File: lib/generator/swaggerMocker.ts [hunk 4]
```diff
@@ -311,10 +366,24 @@ export default class SwaggerMocker {
         ...properties,
         ...this.getProperties(this.getDefSpec(item, visited), visited),
       };
+      this.removeFromSet(item,visited)
```
Comment (raych1) [inline lib/generator/swaggerMocker.ts:369]: Can you add comment of reason to add this remove operation?

File: lib/commands/generate-examples.ts (hunk 1 above)
Comment (jianyexi) [reply lib/commands/generate-examples.ts:29]: Okay

File: lib/generator/swaggerMocker.ts (hunk 3 above)
Comment (raych1) [reply lib/generator/swaggerMocker.ts:157]: Let's rename payloadCache as exampleCache.

File: lib/generator/swaggerMocker.ts (hunk 4 above)
Comment (raych1) [reply lib/generator/swaggerMocker.ts:369]: To resolve a bug.

File: lib/generator/mocker.ts (hunk 2 above)
Comment (jianyexi) [reply lib/generator/mocker.ts:17]: Yes, already replaced almost all, left some to help understand the tool's running progress

File: lib/generator/exampleGenerator.ts [hunk 5]
```diff
@@ -77,61 +113,137 @@ export default class Generator {
...
+    log.info(example);
     const unifiedExample = this.unifyCommonProperty(example);
-
     const newSpec = util.referenceExmInSpec(
       this.specFilePath,
       specItem.path,
       specItem.methodName,
-      `${operationId}_Base`
+      `${operationId}_${rule.exampleNamePostfix}_Base`
     );
```
Comment (raych1) [inline lib/generator/exampleGenerator.ts:196]: Let's call the generated exmaple name as `{operationId}_{rule.postfix}_Gen`.

File: lib/generator/exampleGenerator.ts [hunk 6]
```diff
@@ -77,61 +113,137 @@ export default class Generator {
...
+      specItem = this.getSpecItem(this.spec, operationId);
+      if (!specItem) {
+        console.error(`no specItem for the operation id ${operationId}`);
+        return [];
+      }
+    }
+    const ruleSet: RuleSet = []
+    ruleSet.push({
+      exampleNamePostfix: "Maximum",
+      ruleName: "Maximum"
```
Comment (raych1) [inline lib/generator/exampleGenerator.ts:234]: Can you change it to `MaximumSet` for better understanding.

File: lib/generator/exampleGenerator.ts [hunk 7]
```diff
@@ -77,61 +113,137 @@ export default class Generator {
...
+    });
+    ruleSet.push({
+      exampleNamePostfix: "Minimum",
+      ruleName: "Minimum"
```
Comment (raych1) [inline lib/generator/exampleGenerator.ts:238]: Can you change it to `MinimumSet` for better understanding?

File: lib/generator/exampleRule.ts [hunk 8]
```diff
@@ -1,7 +1,7 @@
 import { CacheItem } from "./exampleCache"
 export interface ExampleRule  {
```
Comment (raych1) [inline lib/generator/exampleRule.ts:2]: Why not use type and make exampleNamePostfix as optional field?

Comment (jianyexi) [reply lib/generator/exampleRule.ts:2]: The two filed are required,  using interface is considering that the rule may be extended in the future

**Comment by raych1** [comment]
Can you update the existed examples which has `_Base` as postfix in the name?

**Comment by raych1** [comment]
Can you fix check failures?

**Review by raych1 — APPROVED** [review]

File: lib/generator/exampleGenerator.ts (hunk 6 above)
Comment (jianyexi) [reply lib/generator/exampleGenerator.ts]: done

**Pull request merged**